pyVor/structures.py

```diff
@@ -0,0 +1,375 @@
+        # try:
+        #     print('here goes {}'.format(self.faces))
+        #     print(str(face))
+        #     self.faces.remove(face)  # ew, linear time.
+        #     print('there went {}'.format(self.faces))
+        #     return [*face.half_facets.values()]
+        # except ValueError:
+        #     print(self.faces)
+        #     raise
+
+    def _facet_pop(self, facet, fsopuwmcd=None):
+        """Push a pin through the facet and out into the twin face, shattering
+        facet.twin.face and also getting rid of facet and facet.twin.
+
+        Or explicitly specify the face to shatter if you like.
+
+        "face_so_other_people_understand_what_my_code_does" is the kwarg.
+        """
+        if not fsopuwmcd:
+            fsopuwmcd = facet.twin.face
+        return [fct for fct in self._face_shatter(fsopuwmcd) if fct !=
+                facet.twin]
+
+    def locate(self, point):
+        """Point location with visibility walk. Straight from my homework."""
+        not_done = True
+        current_face = self._arbitrary_face()
+        while not_done:
+            not_done = False
+            for halffacet in current_face.iter_facets():
+                if halffacet.lineside(point) == -1:
+                    current_face = halffacet.twin.face
+                    not_done = True
+                    break
+        print("Found {} in {}".format(point, current_face.points()))
+        return current_face
+
+    def test_is_delaunay(self):
+        """Make sure every facet is locally delaunay."""
+        for face in self.faces:
+            for halffacet in face.iter_facets():
+                if not halffacet.locally_delaunay():
+                    # print(halffacet)
+                    return False
+        return True
+
+    def face_point_sets(self, homogeneous=False):
+        """Return a set containing a bunch of frozensets of points.
+
+        Of course the frozensets represent the faces of the triangulation.
+        """
+        result = set()
+        # hidden_vertices = frozenset(
+        #   [self.Vertex(point) for point in outer_face_pts(self.dimension())])
+        # for face in self.faces:
+        #     if face.vertices.isdisjoint(hidden_vertices):
+        #         result.add(frozenset([
+        #             vert.point[slice(None) if homogeneous else slice(-1)]
+        #             for vert in face.vertices]))
+        #     else:
+        #         print("It works a little")
+
+        hidden_points = frozenset(outer_face_pts(self.dimension()))
+        for face in self.faces:
+            # [slice(None) if homogeneous else slice(-1)]
+            subresult = frozenset([
+                vert.point for vert in face.vertices])
+            if hidden_points.isdisjoint(subresult):
+                result.add(frozenset([
+                    point[slice(None) if homogeneous else slice(-1)]
+                    for point in subresult]))
+            else:
+                print("It works a little")
```

Comment:
The usefulness of that was that I could tell exactly how many faces met the condition, i.e. contained at least one point at infinity. Way too many did, at the time I added that line. I forget whether I fixed that.